### PR TITLE
Uses `testcontainers` to provide a redis instance for the unit tests

### PR DESCRIPTION
**Pull request by chrisguidry**
We can't assume that a dev would want the unit tests to connect to the locally
running redis at `localhost:6379`, so I'd rather shift that to assuming that the
user has a Docker-compatible system available.  This makes the test suite easier
to reason about and lets us control the redis version with more precision, and
it makes the initial startup experience easier for new devs.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -5,96 +5,90 @@ on:
     branches:
       - main
     tags:
-      - '**'
+      - "**"
```
Comment (chrisguidry) [inline .github/workflows/ci.yml:8]: Sorry, y'all, my VSCode has strong feelings about YAML files

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -5,96 +5,90 @@ on:
...
+      - run: pre-commit run -a --verbose
+        env:
+          SKIP: no-commit-to-branch
 
   docs:
     runs-on: ubuntu-latest
 
     steps:
-    - uses: actions/checkout@v4
+      - uses: actions/checkout@v4
```
Comment (samuelcolvin) [inline .github/workflows/ci.yml:33]: does this need changing now (or at all)? It makes the other changes hear hard to see, and add unnecessary churn.

**Review by samuelcolvin — COMMENTED** [review]
Everything else looks good to me.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (chrisguidry) [reply .github/workflows/ci.yml:33]: Yep sorry about that, I'll get this this tuned back to just the salient changes

**Review by samuelcolvin — APPROVED** [review]

**Comment by samuelcolvin** [comment]
LGTM, merge if you're happy.

**Review by JonasKs — APPROVED** [review]

**Pull request merged**